File: Battleships/java/src/main/java/akuna/battleships/BattleshipClient.java
package akuna.battleships;

import java.io.IOException;
import java.util.Comparator;
import java.util.Map;
import java.util.Random;
import java.util.stream.Collectors;

class BattleshipClient {
    private static final Map<String, Integer> SHIPS = Map.of(
            "aircraftcarrier", 5,
            "battleship", 4,
            "submarine", 3,
            "cruiser", 3,
            "destroyer", 2
    );
    private static final char[] COLS = "abcdefghij".toCharArray();
    private static final char[] ROWS = "0123456789".toCharArray();

    private static String getShipName(char identifier) {
        for (final String shipName : SHIPS.keySet()) {
            if (shipName.charAt(0) == identifier) {
                return shipName;
            }
        }
        return "unknown";
    }

    private final Connection conn;
    private final String name;
    private final Random random;
    private boolean gameOver;

    BattleshipClient(Connection conn, String name) {
        this.conn = conn;
        this.name = name;
        this.random = new Random();
    }

    boolean isGameOver() {
        return gameOver;
    }

    void configure() throws IOException {
        this.conn.send(String.format("id %s", this.name));  // self identify
        this.conn.send("autodump");  // turn on dumping the board after every move
    }

    void placeShips() throws IOException {
        final var col = "a";
        var row = 1;
        final var ships = SHIPS.entrySet().stream()
                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
        for (final var ship : ships) {
            this.conn.send(String.format("place %s %s%d horizontal", ship, col, row));
            this.conn.getLine();
            row++;
        }
    }

    void waitForGo() throws IOException {
        String line;
        do {
            line = this.conn.getLine();
        } while (!line.startsWith("go"));
    }

    void shoot() throws IOException {
        final var col = COLS[this.random.nextInt(COLS.length)];
        final var row = ROWS[this.random.nextInt(ROWS.length)];
        this.conn.send(String.format("shoot %s%s", col, row));
        this.handlePlayerResult(this.conn.getLine());
    }

    void awaitPartnerTurn() throws IOException {
        this.handlePartnerResult(this.conn.getLine());
    }

    private void handlePlayerResult(String result) {
        if (result.startsWith("hit")) {
            System.out.println("Yay!");
        } else if (result.startsWith("miss")) {
            System.out.println("Boo");
        } else if (result.startsWith("sunk")) {
            final var identifier = result.charAt(result.length() - 1);
            System.out.println(String.format("HA! I sunk your %s", BattleshipClient.getShipName(identifier)));
        } else if (result.startsWith("won")) {
            final var identifier = result.charAt(result.length() - 1);
            System.out.println(String.format("HA! I sunk your %s", BattleshipClient.getShipName(identifier)));
            System.out.println("HOORAY!!!");
            this.gameOver = true;
        } else {
            System.out.println(String.format("Unknown result: %s", result));
        }
    }

    private void handlePartnerResult(String result) {
        if (result.startsWith("partner hit")) {
            System.out.println("Darn");
        } else if (result.startsWith("partner miss")) {
            System.out.println("Nah nah nay");
        } else if (result.startsWith("partner sunk")) {
            final var identifier = result.charAt(result.length() - 1);
            System.out.println(String.format("AH! You sunk my %s", BattleshipClient.getShipName(identifier)));
        } else if (result.startsWith("partner won")) {
            final var identifier = result.charAt(result.length() - 1);
            System.out.println(String.format("AH! You sunk my %s", BattleshipClient.getShipName(identifier)));
            System.out.println("drats");
            this.gameOver = true;
        } else {
            System.out.println(String.format("Unknown partner result: %s", result));
        }
    }
}

File: Battleships/csharp/src/Akuna/Battleships/BattleshipClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;

namespace Akuna.Battleships
{
    class BattleshipClient
    {
        private static readonly Dictionary<string, int> SHIPS = new Dictionary<string, int>
        {
            ["aircraftcarrier"] = 5,
            ["battleship"] = 4,
            ["submarine"] = 3,
            ["cruiser"] = 3,
            ["destroyer"] = 2,
        };
        private static readonly char[] COLS = "abcdefghij".ToCharArray();
        private static readonly char[] ROWS = "0123456789".ToCharArray();

        private static String GetShipName(char identifier)
        {
            foreach (String shipName in SHIPS.Keys)
            {
                if (shipName[0] == identifier)
                {
                    return shipName;
                }
            }
            return "unknown";
        }

        private readonly Connection _conn;
        private readonly string _name;
        private readonly Random _random;

        public BattleshipClient(Connection conn, String name)
        {
            _conn = conn;
            _name = name;
            _random = new Random();
        }

        public bool GameOver { get; private set; }

        public void Configure()
        {
            _conn.Send($"id {_name}");  // self identify
            _conn.Send("autodump");  // turn on dumping the board after every move
        }

        public void PlaceShips()
        {
            var col = "a";
            var row = 1;
            var ships = from kvp in SHIPS
                        orderby kvp.Value descending
                        select kvp.Key;
            foreach (var ship in ships)
            {
                _conn.Send($"place {ship} {col}{row} horizontal");
                _conn.GetLine();
                row++;
            }
        }

        public void WaitForGo()
        {
            String line;
            do
            {
                line = _conn.GetLine();
            }
            while (!line.StartsWith("go"));
        }

        public void Shoot()
        {
            var col = COLS[_random.Next(COLS.Length)];
            var row = ROWS[_random.Next(ROWS.Length)];
            _conn.Send($"shoot {col}{row}");
            HandlePlayerResult(_conn.GetLine());
        }

        public void AwaitPartnerTurn()
        {
            HandlePartnerResult(_conn.GetLine());
        }

        private void HandlePlayerResult(String result)
        {
            if (result.StartsWith("hit"))
            {
                Console.WriteLine("Yay!");
            }
            else if (result.StartsWith("miss"))
            {
                Console.WriteLine("Boo");
            }
            else if (result.StartsWith("sunk"))
            {
                var identifier = result[result.Length - 1];
                Console.WriteLine(string.Format("HA! I sunk your {0}", BattleshipClient.GetShipName(identifier)));
            }
            else if (result.StartsWith("won"))
            {
                var identifier = result[result.Length - 1];
                Console.WriteLine(string.Format("HA! I sunk your {0}", BattleshipClient.GetShipName(identifier)));
                Console.WriteLine("HOORAY!!!");
                GameOver = true;
            }
            else
            {
                Console.WriteLine($"Unknown result: {result}");
            }
        }

        private void HandlePartnerResult(String result)
        {
            if (result.StartsWith("partner hit"))
            {
                Console.WriteLine("Darn");
            }
            else if (result.StartsWith("partner miss"))
            {
                Console.WriteLine("Nah nah nay");
            }
            else if (result.StartsWith("partner sunk"))
            {
                var identifier = result[result.Length - 1];
                Console.WriteLine(string.Format("AH! You sunk my {0}", BattleshipClient.GetShipName(identifier)));
            }
            else if (result.StartsWith("partner won"))
            {
                var identifier = result[result.Length - 1];
                Console.WriteLine(string.Format("AH! You sunk my {0}", BattleshipClient.GetShipName(identifier)));
                Console.WriteLine("drats");
                GameOver = true;
            }
            else
            {
                Console.WriteLine($"Unknown partner result: {result}");
            }
        }
    }
}
